x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/BlockingQueryBuilder.java

```diff
@@ -115,4 +119,51 @@ protected int doHashCode() {
     public String getWriteableName() {
         return NAME;
     }
+
+    static class QueryLatch implements Closeable {
+        private volatile CountDownLatch countDownLatch;
+        private Set<Integer> failedShards = new HashSet<>();
+        private int numShardFailures;
+
+        QueryLatch(int numShardFailures) {
+            this.countDownLatch = new CountDownLatch(1);
+            this.numShardFailures = numShardFailures;
+        }
+
+        private void tryBlock(int shardId) throws IOException, InterruptedException {
+            CountDownLatch last = countDownLatch;
+            if (last != null) {
+                last.await();
+            }
+            synchronized (this) {
+                if (failedShards.contains(shardId)) {
+                    throw new IOException("boom");
+                } else if (numShardFailures > 0) {
```

Comment:
can you add a small comment to explain that we have the scenario where the same shard id appears multiple times due to replicas and we want to count that as one failure?

Reply:
also could this block be replaced with the following?

```
if (failedShards.add(shardId)) {
    numShardFailures--;
}
throw new IOException("boom");
```

and the set could be some concurrent impl so that we can remove the synchronized block?

Reply:
I added a comment in https://github.com/elastic/elasticsearch/pull/56026/commits/0304ec6b3b521e2ebe47c40a9c064797203bc021. We could avoid the synchronized block but it feels easier to read and it's not a critical path ?

Reply:
I don't mind the synchronized. On readability, it's debatable I guess, I personally find the `if` based on the `add` result more readable. Up to you.